reflection/serverreflection.go

```diff
@@ -65,8 +65,15 @@ type serverReflectionServer struct {
 
 // Register registers the server reflection service on the given gRPC server.
 func Register(s *grpc.Server) {
-	rpb.RegisterServerReflectionServer(s, &serverReflectionServer{
-		s: s,
+	RegisterWithRegistrar(s, s.GetServiceInfo)
+}
+
+// RegisterWithRegistrar registers the server reflection service with the given registrar.
+// The given function is invoked to determine the services that will be exposed. It is not
+// invoked until the service impl is actually used.
```

Comment:
> It is not invoked until the service impl is actually used.

Is this important for you?

I'm thinking about another option, to take a `map` instead of `func() map`.
It has the advantage that the result will be fixed (instead of dynamic when `queryServices()` is called). But it also loses some flexibility.

Reply:
I think that's fine. I did it this way only so it worked more like the other version, which let you provide the server _before_ it was done accumulating all of the registered services.

Under-the-hood, I think it will need the function and delayed query, just to support the other entry-point.

Reply:
Changed. Let me know what you think.

Reply:
Hmm, right. Sorry, I forgot that `*grpc.Server.GetServiceInfo()` needs to be delayed. Then maybe what you had was better.

But let's wait for @dfawley's opinion on this (he's out this week, and will be back next Monday).

Reply:
And in your use case, are you able to generate a full list of `map[string]grpc.ServiceInfo` before all the services are registered?

Reply:
No, but I can change our framework to register the reflection service _last_ (e.g. wait until the server is actually starting, so all registrations have been accumulated).

Reply:
A different option here is to _not_ add a new function but instead tweak the signature on the existing one to support concrete types other than `*grpc.Server`:

```go
// Registry is a registry of gRPC services. One would typically use a
// *grpc.Server, which implements this interface. But use of an interface
// provides flexibility, such as for testing or use with alternate transport
// implementations.
type Registry interface {
	grpc.ServiceRegistrar
	GetServiceInfo() map[string].grpc.ServiceInfo
}

var _ ServiceRegistry = (*grpc.Server)(nil)

// Register registers the server reflection service on the given gRPC server.
func Register(s ServiceRegistry) {
	rpb.RegisterServerReflectionServer(s, &serverReflectionServer{
		s: s,
	})
}
```

This isn't _strictly_ backwards compatible, yet I noticed the same tactic has already been used to change other references to `*grpc.ClientConn` and `*grpc.Server` to instead use `grpc.ClientConnInterface` and `grpc.ServiceRegistrar`.

Reply:
I was thinking of going one step further and defining an interface in the `grpc` package as well.

```golang
package grpc

type ServerInfoer interface {
  GetServiceInfo() map[string].grpc.ServiceInfo
}
```

And in here we can embed both the interfaces:

```golang
package reflection

type Registrar interface {
	grpc.ServiceRegistrar
	grpc.ServiceInfoer
}
```

But I'm not sure the `ServiceInfoer` interface provides much value outside of the use in the `reflection` package. So, it might not be necessary.

Reply:
I'm throwing yet another idea here:

The interface method can be `GetServiceInfo(string) grpc.ServiceInfo`.

This gives the implementation a chance to intercept the lookup.
While in the old way, the server cannot intercept the map lookup (they cannot overload `[]`).

Reply:
Instead of intercepting map lookup, the simpler interface I described allows a custom impl to intercept the map retrieval -- like modify the map or produce a different/derivative one. Also, your proposed method alone is insufficient -- how does the caller know what string names to use? You'd also have to have a method that provides the list of services exposed.

Finally, if it has the same name (`GetServiceInfo`) then `*grpc.Server` cannot implement it without a backwards incompatible change. So we'd also have to add an adapter type in this package, just to adapt the server to whatever new interface is used.

I personally think sticking with the interface already exposed by `*grpc.Server` would be best.

Reply:
@menghanl or @dfawley, any other thoughts about this?

Reply:
Whoops, didn't see this thread in the "files changed" view.

I agree with your last comment, @jhump.  @easwars what are the benefits of adding this to the grpc package?  That seems counter to Go design ideals - the only reason `ServiceRegistrar` is there is because we didn't want it to be in every single generated code package for...reasons...  I disagreed with them, but went along with it.

Reply:
I think I got confused with supporting the `xds.GRPCServer` and got it wrong here. Yes, no good reason to define it in the `grpc` package.